frontend/src/main/java/org/abs_models/frontend/analyser/ControlFlow.jrag

```diff
@@ -0,0 +1,499 @@
+		}
+		return normalSucc;
+	}
+	eq CaseStmt.getBranch(int i).following() = i + 1 < getNumBranch()
+		? new BitVec<CFGNode>(getBranch(i).getRight()).union(getBranch(i + 1))
+		// We don't need to add the throwTarget, because the compiler generates a pattern for exceptions where needed
+		: new BitVec<CFGNode>(getBranch(i).getRight()).union(following());
+	// Parent CaseStmt, TryCatchFinallyStmt, or null when in recover block
+	inh Stmt CaseBranchStmt.parentStmt();
+	eq CaseBranchStmt.getRight().following() {
+		Stmt parent = parentStmt();
+
+		// If parentStmt is null, we are in a recover block
+		if (parent == null) {
+			return new BitVec<>(exit());
+		}
+
+		if (parent instanceof TryCatchFinallyStmt) {
+			TryCatchFinallyStmt t = (TryCatchFinallyStmt) parent;
+
+			if (t.hasFinally()) {
+				return new BitVec<>(t.getFinally());
+			}
+		}
+
+		return parent.following();
+	}
+
+	eq DieStmt.succ() = new BitVec<>(exit());
+
+	eq DurationStmt.succ() = getMin().canThrow() || getMax().canThrow() ? following().union(throwTarget()) : following();
+
+	eq ExpressionStmt.succ() = getExp().canThrow() ? following().union(throwTarget()) : following();
+
+	eq ForeachStmt.succ() = following().union(getBody()).union(getListExp().canThrow() ? new BitVec<>(throwTarget()) : new BitVec<>());
+	eq ForeachStmt.getBody().following() = new BitVec<>(this);
+
+	eq IfStmt.succ() {
+		BitVec<CFGNode> normalSucc = hasElse()
+			? new BitVec<CFGNode>(getThen()).union(getElse())
+			: following().union(getThen());
+
+		if (getCondition().canThrow()) {
+			return normalSucc.union(throwTarget());
+		}
+		return normalSucc;
+	}
+
+	eq MoveCogToStmt.succ() = getTarget().canThrow() ? following().union(throwTarget()) : following();
+
+	// ReturnStmt does not need extra handling. It has to be the last stmt
+
+	eq ThrowStmt.succ() = new BitVec<CFGNode>(throwTarget());
+
+	eq TryCatchFinallyStmt.succ() = new BitVec<>(getBody());
+	eq TryCatchFinallyStmt.getBody().following() = hasFinally()
+		? new BitVec<CFGNode>(getFinally())
+		: following();
+	eq TryCatchFinallyStmt.getFinally().following() = following().union(throwTarget());
+	eq TryCatchFinallyStmt.getBody().throwTarget() = getCatch(0);
+	eq TryCatchFinallyStmt.getCatch(int i).following() = new BitVec<CFGNode>(getCatch(i).getRight())
+		.union(i + 1 < getNumCatch()
+			// Next catch branch
+			? getCatch(i + 1)
+			// Either finally, or throw again
+			: hasFinally() ? getFinally() : throwTarget());
+
+	eq VarDeclStmt.succ() = getVarDecl().hasInitExp() && getVarDecl().getInitExp().canThrow() 
+		? following().union(throwTarget()) 
+		: following();
+
+	eq WhileStmt.succ() = following().union(getBody()).union(getCondition().canThrow() ? new BitVec<>(throwTarget()) : new BitVec<>());
+	eq WhileStmt.getBody().following() = new BitVec<CFGNode>(this);
+
+    /**
+     * The closest stmt. May be a virtual statement used to model complex expressions such as Let
+     */
+	inh Stmt Exp.parentStmt();
```

Comment:
Is this the same as `Exp.closestParent(Stmt.class)`?  This is already defined in `TreeUtils.jadd`

Reply:
Not Quite. It may also be a virtual statement which is not the actual closest statement in the AST. This preserves the structure while still allowing me to add necessary nodes for LetExp etc.